docs/typescript.md

```diff
@@ -1,14 +1,597 @@
+for (const key of Object.keys(directions) as keyof directions[]) {
+  const direction = directions[key as keyof typeof directions];
+}
+```
+
+##### Type guards can be used to improve type inference and avoid type assertion
+
+```typescript
+function isSomeInterface(x: unknown): x is SomeInterface {
+  return (
+    'name' in x &&
+    typeof x.name === 'string' &&
+    'length' in x &&
+    typeof x.length === 'number'
+  );
+}
+```
+
+🚫 Type assertion
+
+```typescript
+function f(x: SomeInterface | SomeOtherInterface) {
+  console.log((x as SomeInterface).name);
+}
+```
+
+✅ Narrowing with type guard
+
+```typescript
+function f(x: SomeInterface | SomeOtherInterface) {
+  if (isSomeInterface(x)) {
+    // Type of x: 'SomeInterface | SomeOtherInterface'
+    console.log(x.name); // Type of x: 'SomeInterface'. Type of x.name: 'string'.
+  }
+}
+```
+
+```typescript
+const nftMetadataResults = await Promise.allSettled(...);
+
+nftMetadataResults
+  .filter((promise) => promise.status === 'fulfilled')
+  .forEach((elm) =>
+    this.updateNft(
+      elm.value.nft, // Property 'value' does not exist on type 'PromiseRejectedResult'.ts(2339)
+      ...
+    ),
+  );
+```
+
+🚫 Type assertion
+
+```typescript
+(nftMetadataResults.filter(
+    (promise) => promise.status === 'fulfilled',
+  ) as { status: 'fulfilled'; value: NftUpdate }[])
+  .forEach((elm) =>
+    this.updateNft(
+      elm.value.nft,
+      ...
+    ),
+  );
+```
+
+✅ Use a type guard as the predicate for the filter operation, enabling TypeScript to narrow the filtered results to `PromiseFulfilledResult` at the type level
+
+```typescript
+nftMetadataResults.filter(
+    (result): result is PromiseFulfilledResult<NftUpdate> =>
+      result.status === 'fulfilled',
+  )
+  .forEach((elm) =>
+    this.updateNft(
+      elm.value.nft,
+      ...
+    ),
+  );
+```
+
+#### Acceptable usages of `as`
+
+Although `as` is dangerous and discouraged from being used, the following is not intended as an exhaustive list of its valid use cases.
+
+##### `as` is acceptable to use for preventing or fixing `any` usage
+
+Type assertions are unsafe, but they are still always preferred to introducing `any` into the code.
+
+- With type assertions, we still get working intellisense, autocomplete, and other IDE and compiler features using the asserted type.
+- Type assertions also provide an indication of what the author intends or expects the type to be.
+- Often, the compiler will tell us exactly what the target type for an assertion needs to be, enabling us to avoid `as any`.
+
+```typescript
+// Error: Argument of type '"getNftInformation"' is not assignable to parameter of type 'keyof NftController'.ts(2345)
+// 'getNftInformation' is a private method of class 'NftController'
+sinon.stub(nftController, 'getNftInformation');
+```
+
+🚫 `as any`
+
+```typescript
+sinon.stub(nftController, 'getNftInformation' as any);
```

Comment:
It's unclear to me why we'd use `as any` here in the first place.

Maybe we can remove this guideline as well? This is another case where the guideline is about dealing with bad TypeScript, rather than legitimate scenarios where `as` is needed

Reply:
I have similar thoughts. I understand that people may have questions about when it is okay to use `as`, and because of this, perhaps this section is intended to be informational rather than advisory. But, I worry that people may cite this section as evidence that it is okay to do this, especially since we say that this is "acceptable", and to avoid this I feel like we need to be more guarded in the language.

I think it would be more helpful to make this a proper guideline and then to use specific language which clarifies _when_ it is okay. Perhaps instead of "`as` is acceptable to use for preventing or fixing `any` usage" we say "Prefer type assertions over disabling type-checking" or something like that?

Besides this, I'm curious if there is a better example, since we don't want people testing private methods to begin with.

Reply:
I am okay with leaving the example for now and improving it later (given that it _is_ based on a real-world example), but I'm still curious whether it makes sense to rename the header.